### PR TITLE
Auth on server must be non-interactive. Auth on client should be interactive.

### LINKED ISSUE #9659: Auth in Visual Studio should not always be interactive
When running in VS.Oe, NuGet should not ask credential providers to be interactive.

### PR DESCRIPTION
**Pull request by rrelyea**
## Bug

Fixes: https://github.com/NuGet/Home/issues/9659
Regression: No

## Fix

Check VSOe state to determine nonInteractive flag for credential providers.

## Testing/Validation

Tests Added: No
Reason for not adding tests: will add to manual test pass, once working end to end. 
Validation: none yet...need to get fixes from 3 components.

### CODE REVIEW

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/DefaultVSCredentialServiceProvider.cs [hunk 1]
```diff
@@ -87,7 +87,10 @@ public async Task<NuGet.Configuration.ICredentialService> GetCredentialServiceAs
             }
 
             // Initialize the credential service.
-            var credentialService = new CredentialService(new AsyncLazy<IEnumerable<ICredentialProvider>>(() => System.Threading.Tasks.Task.FromResult((IEnumerable<ICredentialProvider>)credentialProviders)), nonInteractive: false, handlesDefaultCredentials: PreviewFeatureSettings.DefaultCredentialsAfterCredentialProviders);
+            var credentialService = new CredentialService(
+                new AsyncLazy<IEnumerable<ICredentialProvider>>(() => System.Threading.Tasks.Task.FromResult((IEnumerable<ICredentialProvider>)credentialProviders)),
+                nonInteractive: KnownUIContexts.CloudEnvironmentConnectedContext.IsActive, // cannot only interact when cloudconnection is not active.
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/DefaultVSCredentialServiceProvider.cs]: I don't know if this is the right thing to check. 
I think this is only true when there's a client connected to the server. (maybe only client side)
You might be looking for something like https://github.com/NuGet/NuGet.Client/blob/e3c62df98bec2b6232c0cec93cb60b5fa1d9aa95/src/NuGet.Clients/NuGet.Console/OutputConsole/OutputConsoleProvider.cs#L123-L130 instead. 

I could be wrong though.

Comment (rrelyea) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/DefaultVSCredentialServiceProvider.cs]: likely we'll need that several places in our code. 
options: 
1) call VS method:  Microsoft.VisualStudio.Cascade.CommonUtility.IsVSOnlineServerAsync()
2) move method from OutputConsoleProvider to NuGet.Common somewhere (or where do we keep things like this?)
3) copy the code

Thoughts?

I lean towards num 1 if layering makes sense...or num 2

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/DefaultVSCredentialServiceProvider.cs]: NuGet.VisualStudio.Common likely. 

1 or 2 are both fine I think. You can use 1 if we have access to it. For 2, I think you likely want NuGet.VisualStudio.Common.

Comment (rrelyea) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/DefaultVSCredentialServiceProvider.cs]: i'm prototyping num 2. moving our helper to NuGet.VisualStudio.Common.dll - NuGet.VisualStudio.VisualStudioContextHelper class - IsVSOnlineServerAsync method moved from OutputConsoleProvider.
Thoughts?

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/DefaultVSCredentialServiceProvider.cs]: I think that's fine.

**Review by nkolev92 — APPROVED** [review]

File: src/NuGet.Clients/NuGet.VisualStudio.Common/VisualStudioContextHelper.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace NuGet.VisualStudio
+{
+    public static class VisualStudioContextHelper
+    {
+        public static async Task<bool> IsInServerModeAsync(CancellationToken token)
+        {
+            await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(token);
+            IVsShell shell = await ServiceLocator.GetGlobalServiceAsync<SVsShell, IVsShell>();
```
Comment (sbanni) [inline src/NuGet.Clients/NuGet.VisualStudio.Common/VisualStudioContextHelper.cs:16]: Should check shell for null, consider turning on #nullable enable for this file.

**Review by sbanni — APPROVED** [review]

**Pull request merged**